Minimum over lap size

jm-c opened the pull request:
## What changes does this PR introduce?
Clarify minimum overlap-size needed for main model.
 1) fix few loop range to allow to run with OLx,y=1 when momStepping=F ;
 2) add check and stop if momStepping=T and overlap-size is smaller than 2.

## What is the current behaviour? 
minimum overlap-size is checked for to satisfy tracer advection requirement (in gad_check.F)
and  biharmonic viscosity. However, there is no check related to other part of the main model,
specially whether one could use  overlap-size of just 1  

## What is the new behaviour 
- fix few loop range in main model src code (ini_cg2d.F, uv exch) to enable to run with overlap-size = 1,
  as long as momStepping=F.
- also add a check and stop for minimum overlap-size.

## Does this PR introduce a breaking change? 
no

## Other information:
The few minor changes here allow, e.g, to use (safely) OLx,y=1 for offline and tracer advection problems.

## Suggested addition to `tag-index`
o model/src: 
  - fix few loop range to allow to run with OLx,y=1 when momStepping=F ;
  - add check and stop if momStepping=T and overlap-size is smaller than 2.

mjlosch commented on model/src/config_check.F:
```diff
@@ -463,6 +463,16 @@ SUBROUTINE CONFIG_CHECK( myThid )
 C--   Check parameter consistency :
 
+      IF ( ( OLx.LT.2 .OR. OLy.LT.2 ) .AND. momStepping ) THEN
+C-    Current algorithm (solve_for_pressure, RHS) prevent to use
```
prevent -> prevents

jm-c replied:
> should there be a check for OLx/y < 1? 
I could add an unconditional check & stop for this but it might not be very useful because the model might stop earlier when accessing "out of bounds" array element.

> Also for viscAh, etc. > 0 in anology to viscA4>0 and OLx/y<3?
This is covered by this current PR (but not in current master yet) which check for "momStepping".

> I find the comment in config_check a bit misleading, because it implies that it still doesn't work ("but some loop range in few momentum S/R are not carefully set to support this"). I would write instead (if that's correct):
> 
> ```
> C     The current algorithm (solve_for_pressure, RHS) prevents to use
> C     momAdvection or viscAh > 0 with OLx,y=1 ; we can still allow to
> C     use OLx,y=1 when momAdvection=F and viscAh=0. Note that some loop
> C     ranges in few momentum S/R are not carefully set to support this.
> ```
Regarding this point, I would prefer not to use an affirmative "we can still allow" for something that currently does not work, not because of the algorithm itself but because of how it's coded. @mjlosch I am fixing the typo + try to clarify a bit and let me know if it's still not clear.

mjlosch replied:
@jm-c 
I am not sure if this is worth the trouble, but this is way I read the code + comments in the current form: We have an algorithm that we cannot use for OLx/y<2 and momStepping=T, that's why we stop here. Using it with OLx/y=1 and momAdvection=F works, but some loop ranges are not adjusted properly so that it doesn't work, but we still allow it. I think you mean this:
```
C     The current algorithm (solve_for_pressure, RHS) prevents to use
C     momAdvection or viscAh > 0 with OLx,y=1, because some loop
C     ranges in few momentum S/R are not carefully set to support this.
C     It is, however, possible to use OLx,y=1 when momAdvection=F and 
C     viscAh=0.
```

Again, only if you think that this is important.

jm-c replied:
@mjlosch Thanks, now I see where the confusion is: I always have in mind that if momStepping=F then momAdvection=F, because of line 56 in set_parms.F, so I was taking the point of view of what the model does
(and as reported in config_summary.F) whereas you were considering what the user is setting up from parameter file "data".

mjlosch reviewed (commented):
should there be a check for OLx/y < 1? Also for viscAh, etc. > 0 in anology to viscA4>0 and OLx/y<3?

I find the comment in config_check a bit misleading, because it implies that it still doesn't work ("but some loop range in few momentum S/R are not carefully set to support this"). I would write instead (if that's correct):
```
C     The current algorithm (solve_for_pressure, RHS) prevents to use
C     momAdvection or viscAh > 0 with OLx,y=1 ; we can still allow to
C     use OLx,y=1 when momAdvection=F and viscAh=0. Note that some loop
C     ranges in few momentum S/R are not carefully set to support this.
```

jm-c commented:
I added an other check & stop if OLx,OLy < 1 ; I put this in set_parms.F since the overlap are used in some parts of INI_GRID and in config_check.F it would have been too late (I added a comment in the code for this).

jm-c commented:
Unless someone disagree, going to merge this PR later today or tomorrow.
